0005-R-pluggable-components-injector.md

```diff
@@ -0,0 +1,298 @@
+
+Example:
+
+```yaml
+apiVersion: dapr.io/v1alpha1
+kind: Component
+metadata:
+  name: my-component
+  annotations:
+    dapr.io/component-container-image: "component:v1.0.0"
+spec:
+  type: state.my-componentƒ
+  version: v1
+  metadata: []
+```
+
+Optionally you can mount volumes and add env variables into the containers by using the `dapr.io/component-container-volume-mounts(-rw)` and `dapr.io/component-container-env` annotations.
+
+```yaml
+apiVersion: dapr.io/v1alpha1
+kind: Component
+metadata:
+  name: my-component
+  annotations:
+    dapr.io/component-container-image: "component:v1.0.0"
+    dapr.io/component-container-volume-mounts: "volume-name:/volume-path,volume-name-2:/volume-path-2" # read-only, "$VOLUME_NAME:$VOLUME_PATH,$VOLUME_NAME_2:$VOLUME_PATH2"
+    dapr.io/component-container-volume-mounts-rw: "volume-name-rw:/volume-path-rw,volume-name-2-rw:/volume-path-2-rw" # read-write "$VOLUME_NAME:$VOLUME_PATH,$VOLUME_NAME_2:$VOLUME_PATH2"
+    dapr.io/component-container-env: "env-var=env-var-value,env-var-2=env-var-value-2" #optional "$ENV_NAME=$ENV_VALUE,$ENV_NAME_2=$ENV_VALUE_2"
+spec:
+  type: state.my-component
+  version: v1
+  metadata: []
+```
+
+By default the injector creates undeclared volumes as `emptyDir` volumes, if you want a different volume type you should declare it by yourself in your pods.
+
+#### Pod annotations
+
+In order to allow users to turn off the component injector for their pod, a new annotation will be available, similar to the one that we have for enabling dapr: `dapr.io/inject-pluggable-components:"true"`. Let's rewrite the previous examples using the injected pluggable components feature, it would be something like:
+
+The apps deployments/pods:
+
+```yaml
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+name: app
+labels:
+  app: app
+spec:
+replicas: 1
+selector:
+  matchLabels:
+    app: app
+template:
+  metadata:
+    labels:
+      app: app
+    annotations:
+      dapr.io/inject-pluggable-components: "true"
+      dapr.io/app-id: "my-app"
+      dapr.io/enabled: "true"
+  spec:
+    containers:
+      - name: my-app
+        image: my-app-image:latest
+---
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+name: app-2
+labels:
+  app: app-2
+spec:
+replicas: 1
+selector:
+  matchLabels:
+    app: app-2
+template:
+  metadata:
+    labels:
+      app: app-2
+    annotations:
+      dapr.io/inject-pluggable-components: "true"
+      dapr.io/app-id: "my-app-2"
+      dapr.io/enabled: "true"
+  spec:
+    containers:
+      - name: my-app-2
+        image: my-app-2-image:latest
+```
+
+And the component spec:
+
+```yaml
+apiVersion: dapr.io/v1alpha1
+kind: Component
+metadata:
+  name: my-component
+  annotations:
+    dapr.io/component-container-image: "component:v1.0.0"
```

Comment:
How is this going to get injected into the pod? would the sidecar injector now need to get a list of all Dapr components in the namespace and create the corresponding containers in the pod?

Reply:
There are few alternatives,

Ideally, we should guarantee that whatever the component list will be fetched from, it should be consistent with the runtime component list to avoid racing conditions between the former and the latter, e.g in case of a pluggable component is configured between the injector patch and the pod start. 

Even considering that racing condition, the reconciling k8s philosophy will help to reach the desirable state somehow.

In the case of consistency being a `MUST HAVE` I see two phases to achieve that

phase 1: Rely on k8s

1. Use the daprClient to fetch the component list and apply the filter by scope logic 
* Advantage:
    - do not depend on operator to work
* Disadvantage:
    - have to "duplicate" the logic to filter by scope given the app name
2. Fetch from operator API
* Advantage:
    - do not have to "duplicate" the logic to filter by scope given the app name
* Disadvantage:
    - depend on operator to work

phase 2: Consistency guaranteed

I don't have a clear path here, it could be something like a `components version` that allow to fetch a specific version of the dapr configuration state, or something like we've discussed about having a init container that populates the components directory based on a base64-encoded components yaml.


----

What do you think @yaron2  ?

Reply:
> it should be consistent with the runtime component list

What's the "runtime component list"? components are stored on on etcd in Kubernetes and are basically a shared global state. What race condition do you mean here?

> Use the daprClient to fetch the component list and apply the filter by scope logic

This should be the way to go, coupling the injector to the operator is extremely problematic for a number of reasons. I'm fine with this approach, the think the major disadvantage here is a `List` call to the k8s API server at scale. This is, frankly, the thing that worries me the most.

> I don't have a clear path here, it could be something like a components version that allow to fetch a specific version of the dapr configuration state, or something like we've discussed about having a init container that populates the components directory based on a base64-encoded components yaml.

Both paths here are a non-starter for me, but that's not relevant for now.

Reply:
> What's the "runtime component list"? components are stored on on etcd in Kubernetes and are basically a shared global state. What race condition do you mean here?

I meant this list => https://github.com/dapr/dapr/blob/master/pkg/runtime/runtime.go#L2372

Try to imagine a scenario where the injector asks for all components in the current namespace in a timestamp T1, shortly after (before the target application pod gets running) the user adds a new pluggable component (i.e a new component yaml) T2, in a timestamp T3 the pod is ready to run and it will retrieve the list again from the operator using the [loader](https://github.com/dapr/dapr/blob/master/pkg/runtime/runtime.go#L2372)

The list from runtime will be different from the list that the injector used when patching the pod.

As I said It could lead in two different situations:

1) inject a useless pod in case of T2 being a deletion of a component
2) do not inject a required pod that might cause panics